src/task2/graph_drawer.py

```diff
@@ -174,3 +167,70 @@ def setFontSize(self, font_size):
         font.setPointSize(font_size)
         self.setFont(font)
         self.resize(self.sizeHint())
+
+
+class Town(QWidget):
+    def __init__(self, idx, radius, parent=None, font_size=None):
+        super().__init__(parent)
+
+        self.town_idx = idx
+        self.town_type = 0
+        self.font_coefficient = 1.5
+        self.idx_label = CustomLabel(str(idx), self)
+        self._outer_radius = radius
+        self._inner_radius = int(radius * 0.8)  # ? property
+        self._font_size = font_size or int(1 / len(str(idx)) * self._inner_radius * self.font_coefficient)
+        self._scale = 0
+        self.thickness = 2
+        self.setTownInfo()
+        self.show()
+
+    @property
+    def outer_radius(self):
+        return int(self.scale * self._outer_radius) + 1
+
+    @property
+    def inner_radius(self):
+        return int(self.scale * self._inner_radius) + 1
+
+    @property
+    def font_size(self):
+        return int(self.scale * self._font_size) + 1
+
+    @property
+    def scale(self):
+        return self._scale
+
+    @scale.setter
+    def scale(self, value):
+        self._scale = value
+        self.resize((self.outer_radius + self.thickness + 1) * 2, (self.outer_radius + self.thickness + 1) * 2)
+
+    def paintEvent(self, e):
+        inner_radius = self.inner_radius
+        outer_radius = self.outer_radius
+        font_size = self.font_size
+
+        painter = QPainter()
+        painter.begin(self)
+
+        painter.setPen(QPen(Qt.black, self.thickness))
+
+        painter.drawEllipse(self.thickness, self.thickness, 2 * outer_radius, 2 * outer_radius)
+
+        _, color = buildings_to_info[self.town_type]
+        painter.setBrush(color)
+
+        diff = outer_radius - inner_radius
+        painter.drawEllipse(diff + self.thickness, diff + self.thickness, 2 * inner_radius, 2 * inner_radius)
+
+        self.idx_label.setFontSize(font_size)
+        self.idx_label.move(outer_radius - self.idx_label.width() // 2, outer_radius - self.idx_label.height() // 2)
+
+        painter.end()
+
+    def setTownInfo(self, idx=None, town_type=None,):
+        self.town_idx = idx or self.town_idx
+        self.town_type = town_type or self.town_type
+
+        self.setToolTip(f'Idx: {self.town_idx}\nType: {buildings_to_info[self.town_type][0]}')
```

Comment:
do some structure for saving town info ?
make class and override `__str__` method

Reply:
We need a static class to store all the information requested from the server (levels, user information, maybe something else) because this data is requested from different classes and will be wastefully stored in each class.

Reply:
@Madiy, check now
I thick, that this is good solution, because I store only reference to object, which save information for point
And I can change this object from any place
Also all logic in App class, we can create object there, and pass them to class, which need them

Reply:
Looks good